File: backend/internal/entities/student.go
package entities

import "time"

// StudentProfile represents a student's profile information
type StudentProfile struct {
	UserID        int       `json:"user_id"`
	CurrentStreak int       `json:"current_streak"`
	LongestStreak int       `json:"longest_streak"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`

	// Related entities (not in the database)
	User      *User          `json:"user,omitempty"`
	Languages []UserLanguage `json:"languages,omitempty"`
	Interests []UserInterest `json:"interests,omitempty"`
	Goals     []UserGoal     `json:"goals,omitempty"`
}

// StudentRegistrationRequest represents the data needed to register as a student
type StudentRegistrationRequest struct {
	// Basic user registration data
	Username     string `json:"username" validate:"required"`
	PasswordHash string `json:"password_hash" validate:"required"`
	Email        string `json:"email" validate:"required,email"`

	// Student-specific data (optional)
	Languages []UserLanguageUpdate `json:"languages,omitempty"`
	Interests []int                `json:"interests,omitempty"`
	Goals     []int                `json:"goals,omitempty"`
}

// StudentUpdateRequest represents the data needed to update a student's profile
type StudentUpdateRequest struct {
	ProfilePictureURL *string              `json:"profile_picture_url,omitempty"`
	Languages         []UserLanguageUpdate `json:"languages,omitempty"`
	Interests         []int                `json:"interests,omitempty"`
	Goals             []int                `json:"goals,omitempty"`
}

File: frontend/src/types/student.ts
import { BaseEntity } from './common';
import { User } from './user';
import { UserGoal } from './interest-goal';
import { UserInterest } from './interest-goal';
import { UserLanguage, UserLanguageUpdate } from './language';

/**
 * Student-related types
 */

// Student profile
export interface StudentProfile extends BaseEntity {
  user_id: number;
  current_streak: number;
  longest_streak: number;
  user?: User;
  languages?: UserLanguage[];
  interests?: UserInterest[];
  goals?: UserGoal[];
}

// Student registration
export interface StudentRegistrationRequest {
  username: string;
  password: string; // For frontend usability
  email: string;
  languages?: UserLanguageUpdate[];
  interests?: number[];
  goals?: number[];
}

// Student update request
export interface StudentUpdateRequest {
  profile_picture_url?: string;
  languages?: UserLanguageUpdate[];
  interests?: number[];
  goals?: number[];
} 